### PR TITLE
Non-record: Annealed Muon 1.58-bit Ternary — val_bpb 1.2196 (8xH100 SXM)

### PR DESCRIPTION
**Pull request by DushyantChetiwal**
## Summary

- **val_bpb: 1.2196** | **14.86 MB** | 8xH100 SXM, 595s
- Training-time 1.58-bit ternary quantization (AnnealedBitLinear) with Muon optimizer
- 800d / 10L / 8 heads / 8 KV heads / MLP 4x relu^2 / 524k batch tokens
- Base-3 packing (5 ternary values/byte) + zstd-22 compression
- Roundtrip val_bpb: 1.2249

## Context

Fair warning: I'm a complete beginner to LLM training -- I started about two weeks ago and this challenge was how I chose to learn. A lot of what's in here was figured out through trial and error rather than deep prior knowledge, so apologies in advance if some of the design choices look naive or if I've reinvented wheels poorly. That said, I learned an enormous amount through the process and wanted to submit what I have, rough edges and all.

## Approach

Explores quantization-aware training with ternary weights ({-1, 0, +1}) as an alternative to the dominant post-training GPTQ int6 approach. Weights are stored as full-precision latents during training and progressively annealed toward hard ternary quantization via a phi-exponent schedule.

Key architectural features: U-Net skip connections, XSA (cross-sequence attention), BigramHash embedding (2048 buckets), SmearGate, and hold-cosine LR schedule (hold=0.70).

Much of the development was empirical -- running ablations, seeing what broke, and trying to understand why after the fact. Some experiments that didn't work out but were educational:
- **Ternary Bucket Voting (TBV)**: a weight averaging algorithm I designed to respect ternary structure by voting on majority ternary class per weight. Sounded reasonable on paper, failed catastrophically in practice (val_bpb 2.2475) -- high LR + averaging produces weights that collapse to zero.
- **EMA/SWA**: standard weight averaging destroys ternary structure through sign cancellation. I had to learn this the hard way before finding the ternary SOTA's RESULTS.md confirming the same.
- **Width scaling (D=800)**: marginal gain over D=768, offset by halved step count from slower per-step compute.
- **LR schedule tuning**: reducing the hold phase from 0.70 to 0.30 hurt, which was counterintuitive to me at the time.

## Why non-record

Ternary weights ({-1, 0, +1}) pack more parameters per MB than int6, but each individual weight is much less precise -- and that noise adds up through the layers. This is a hard limit of the representation that no amount of training tricks could fix. On top of that, I only realized late in the process that the 1024-token vocabulary was costing a huge amount of BPB compared to the 8192-token vocab used by the ternary SOTA (1.1570 BPB) -- a gap that has nothing to do with the model architecture at all.

## Files

- `train_gpt.py` -- full training script
- `submission.json` -- metadata
- `train_log.txt` -- 8xH100 SXM training log
- `README.md` -- detailed description

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: Annealed Muon 1.58-bit Ternary — val_bpb 1.2196 (8xH100 SXM)

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

## Analysis

### PR Overview
PR #1273: "Annealed-Muon 1.58-bit 8KV" — a pure neural model using 1.58-bit quantization (AnnealedBitLinear), Muon+ optimizer, grouped parameter tying, U-Net skips, and XSA attention. Head SHA: 5b6f4656ddda1d3993e4258f5a5a162015330a78.

### ILLEGAL N-gram Family Bug Check
The `BigramHashEmbedding` class (lines 412–436) computes:
```python
out[..., 1:] = torch.bitwise_xor(36313 * t[..., 1:], 27191 * t[..., :-1]) % mod
```
This hashes `(token[t], token[t-1])` pairs — both are **input** tokens from `input_ids`. The hash key is formed exclusively from prior context tokens; no target token is XOR'd into the key. The bigram embedding is applied in `forward()` (line 572) as an additive offset to the token embedding, using only `input_ids`, before any target is seen. **No ILLEGAL n-gram family bug.**

### ILLEGAL Pre-Quant TTT Check
No test-time training loop is present. The entire training loop (lines 800–868) reads from `train_loader` (train split files). `val_tokens` is only ever passed to `eval_val()` (lines 793–794 and 940–942), which runs entirely under `torch.no_grad()` (line 233). There are zero calls to `backward()`, `optimizer.step()`, or `model.train()` that involve `val_tokens`. **No Pre-Quant TTT.**

### Score-First TTT Check
No TTT of any kind is implemented. **Not applicable.**

### Scored-Region SLOT Check
No held-out scored region or SLOT usage detected. **Not applicable.**

### Summary
This is a straightforward pure neural submission. The BigramHashEmbedding is a standard learned context-dependent input embedding (prior token pair → additive bias), operating only on input tokens. All validation is inference-only. Training uses only train-split data. No forbidden algorithmic patterns detected.

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
